feat(clean): add clean_headers function

atol opened the pull request:
# Description

Implements the `clean_headers()` function to clean column names. Design details are set out in #447.

# How Has This Been Tested?

See snapshots below.

# Snapshots:

## Functionality

<img width="565" alt="clean_headers" src="https://user-images.githubusercontent.com/55072906/106370604-6892a680-6310-11eb-8a02-535563333b87.png">

## Specify case style

Users can specify the case style using the `case` parameter:

<img width="546" alt="styles_1" src="https://user-images.githubusercontent.com/55072906/106370614-7a744980-6310-11eb-8d95-2a4a6f731917.png">

<img width="602" alt="styles_2" src="https://user-images.githubusercontent.com/55072906/106370616-7cd6a380-6310-11eb-95c0-d3405324ba7c.png">

<img width="604" alt="styles_3" src="https://user-images.githubusercontent.com/55072906/106370617-7f38fd80-6310-11eb-8127-4ee77088b752.png">

## Replace values

Users can specify values to replace in the column names by passing in a dictionary to the `replace` parameter:

<img width="578" alt="replace" src="https://user-images.githubusercontent.com/55072906/106370635-abed1500-6310-11eb-9843-654a6b7b1af1.png">

## Remove accents

Accents are automatically removed from column names:

<img width="518" alt="accents_1" src="https://user-images.githubusercontent.com/55072906/106370627-9a0b7200-6310-11eb-8d7f-305d90beaff3.png">

Removal of accents can be suppressed by setting `remove_accents=False`:

<img width="518" alt="accents_2" src="https://user-images.githubusercontent.com/55072906/106370629-9d9ef900-6310-11eb-94b2-1306a04567ed.png">

## Rename duplicated column names

Column names duplicated by calling `clean_headers()` are automatically renamed by appending a number to the end of the duplicated column name(s):

<img width="518" alt="duplicate_1" src="https://user-images.githubusercontent.com/55072906/106370649-ce7f2e00-6310-11eb-8e60-e220c2fe54e0.png">

The suffix used to append the number is inferred from the `case` type:

<img width="519" alt="duplicate_2" src="https://user-images.githubusercontent.com/55072906/106370652-da6af000-6310-11eb-8803-8929e5d89aa0.png">

## Real World Data

An example of `clean_headers()` called on the [FIFA 21 messy, raw dataset for cleaning/ exploring](https://www.kaggle.com/yagunnersya/fifa-21-messy-raw-dataset-for-cleaning-exploring) dataset:

<img width="802" alt="real_world_1" src="https://user-images.githubusercontent.com/55072906/106370678-43526800-6311-11eb-895e-c30e86999c15.png">

<img width="803" alt="real_world_2" src="https://user-images.githubusercontent.com/55072906/106370681-48afb280-6311-11eb-926d-9ce443120ada.png">

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have already squashed the commits and made the commit message conform to the project standard.
- [ ] I have already marked the commit with "BREAKING CHANGE" or "Fixes #" if needed.
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

brandonlockhart commented on dataprep/clean/clean_headers.py:
```diff
@@ -0,0 +1,273 @@
+
+    if case == "snake":
+        name = "_".join(words).lower()
+    if case == "kebab":
```
Should this be `elif` instead of repeating `if` statements?

brandonlockhart commented on dataprep/clean/clean_headers.py:
```diff
@@ -0,0 +1,273 @@
+        )
+
+    # Convert to dask
+    df = to_dask(df)
```
I recommend not going converting to dask. I think the API to access the columns for a Dask and pandas DataFrame is `df.columns`, so I don't think you will need to consider whether a Dask or pandas df is passed.

brandonlockhart commented on dataprep/clean/clean_headers.py:
```diff
@@ -0,0 +1,273 @@
+    # Count null column values
+    stats["null"] = len([col for col in list(df.columns) if col in NULL_VALUES])
+
+    with ProgressBar(minimum=1, disable=not progress):
```
I think you can remove `dask.compute(df)` here, and also the progress bar parameter from `clean_headers`.

brandonlockhart reviewed (commented):
Very nice PR Alice! I like how you structured the code and your string manipulation looks good to me, and thanks for following the docstring/coding guidelines carefully. I left some small comments below.

I'm wondering about how we're handling null headers. What do you think about creating a default value, eg 'header', and replacing any null headers with the value 'header'? So if we have three null column headers and `case='snake'`, the null column headers would be replaced with 'header', 'header_2', 'header_3'? (I think this is how you handle duplicate headers) 

Let me know what you think.

atol commented:
Thanks for your comments, Brandon! I agree with your suggestion on how to handle null headers and have implemented it accordingly. Since we now rename null headers, it seems unnecessary to keep track of null values. As such, I have removed that statistic from the cleaning report as well.

<img width="400" alt="null_headers" src="https://user-images.githubusercontent.com/55072906/106843157-38257200-665a-11eb-9c8a-b9c8af7a7b02.png">

brandonlockhart commented on dataprep/clean/clean_headers.py:
```diff
@@ -0,0 +1,246 @@
+"""
+Implement clean_headers function
```
Can you please take a look at point 1. (Module Docstring) [here](https://github.com/sfu-db/dataprep/wiki/Docstring-Best-Practice#dataprepclean-docstring-recommendations). I'm wondering if we could write a clearer brief description of the functionality in this module, perhaps something like "Clean and standardize column headers." (or whatever you think is best)

brandonlockhart reviewed (changes requested):
Thanks for making the revisions, Alice! I just left one more comment.

Also, can you please follow the instructions [here](https://github.com/sfu-db/dataprep/wiki/Docstring-Best-Practice#auto-generating-the-api-reference) and add the modified file to this PR so that your docstrings will be added to the documentation in the next release?

The function works great from my testing though; so after these two changes, I think it can be merged!

atol commented:
Sorry about that! I wasn't sure whether to include the changes to `dataprep.clean.rst` in this PR or a later one for the docs. I've added the modified file and updated the Module Docstring.

brandonlockhart commented:
No problem! I think `dataprep.clean.rst` can be added at this stage since the docstrings are ready. And if there's a release before the user guide is added, the user can still learn about this function from the API reference.
